From d7fcc5b770969fe9d27925fe8b6338b3de88b3d7 Mon Sep 17 00:00:00 2001
From: Bai Ping <ping.bai@nxp.com>
Date: Wed, 8 Nov 2017 17:58:00 +0800
Subject: [PATCH 2789/5242] MLK-16804-06 driver: soc: Optimize the DDR
 frequency in audio playback case

commit  ada612ab471a0ed7c39106f1292c0bc6d93d923e from
https://source.codeaurora.org/external/imx/linux-imx.git

If audio device is the only that access to ddr memory, the DDR
frequency can be reduce to 25MHz to save power. when DDR run in
25MHz frequency, the memory bandwidth is about 66MB/s, it can
meet the performance requirement for audio only case.

Signed-off-by: Bai Ping <ping.bai@nxp.com>
Reviewed-by: Anson Huang <anson.huang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/soc/imx/busfreq-imx8mq.c |   16 +++++++++++-----
 1 file changed, 11 insertions(+), 5 deletions(-)

diff --git a/drivers/soc/imx/busfreq-imx8mq.c b/drivers/soc/imx/busfreq-imx8mq.c
index f4d970d..4054c37c 100644
--- a/drivers/soc/imx/busfreq-imx8mq.c
+++ b/drivers/soc/imx/busfreq-imx8mq.c
@@ -102,14 +102,19 @@ static void reduce_bus_freq(void)
 	clk_prepare_enable(sys1_pll_40m);
 	clk_prepare_enable(dram_alt_root);
 
+	/*
+	 * below piece of code has some redundant part, keep
+	 * it at present, we may need update the audio freq
+	 * in the future if needed.
+	 */
 	if (audio_bus_count) {
-		clk_prepare_enable(sys1_pll_400m);
+		clk_prepare_enable(sys1_pll_100m);
 
-		update_bus_freq(AUDIO_FREQ_400MTS);
+		update_bus_freq(LOW_BUS_FREQ_100MTS);
 
 		/* correct the clock tree info */
-		clk_disable_unprepare(sys1_pll_400m);
-		clk_set_parent(dram_alt_src, sys1_pll_400m);
+		clk_disable_unprepare(sys1_pll_100m);
+		clk_set_parent(dram_alt_src, sys1_pll_100m);
 		clk_set_parent(dram_core_clk, dram_alt_root);
 		clk_set_parent(dram_apb_src, sys1_pll_40m);
 		clk_set_rate(dram_apb_pre_div, 20000000);
@@ -128,6 +133,7 @@ static void reduce_bus_freq(void)
 		clk_set_parent(dram_core_clk, dram_alt_root);
 		clk_set_parent(dram_apb_src, sys1_pll_40m);
 		clk_set_rate(dram_apb_pre_div, 20000000);
+		clk_prepare_enable(sys1_pll_400m);
 
 		low_bus_freq_mode = 1;
 		audio_bus_freq_mode = 0;
@@ -138,7 +144,7 @@ static void reduce_bus_freq(void)
 	clk_disable_unprepare(dram_alt_root);
 
 	if (audio_bus_freq_mode)
-		printk(KERN_DEBUG "ddrc freq set to audio mode: 100MHz\n");
+		printk(KERN_DEBUG "ddrc freq set to audio mode: 25MHz\n");
 	if (low_bus_freq_mode)
 		printk(KERN_DEBUG "ddrc freq set to low bus mode: 25MHz\n");
 }
-- 
1.7.9.5

